src/main/java/net/neoforged/neoforge/registries/DataMapLoader.java

```diff
@@ -0,0 +1,178 @@
+import net.neoforged.neoforge.registries.datamaps.DataMapFile;
+import net.neoforged.neoforge.registries.datamaps.DataMapType;
+import net.neoforged.neoforge.registries.datamaps.DataMapValueMerger;
+import org.slf4j.Logger;
+
+@SuppressWarnings({ "rawtypes", "unchecked" })
+public class DataMapLoader implements PreparableReloadListener {
+    private static final Logger LOGGER = LogUtils.getLogger();
+    public static final String PATH = "data_maps";
+    private Map<ResourceKey<? extends Registry<?>>, LoadResult<?>> results;
+    private final ICondition.IContext conditionContext;
+    private final RegistryAccess registryAccess;
+
+    public DataMapLoader(ICondition.IContext conditionContext, RegistryAccess registryAccess) {
+        this.conditionContext = conditionContext;
+        this.registryAccess = registryAccess;
+    }
+
+    @Override
+    public CompletableFuture<Void> reload(PreparationBarrier pPreparationBarrier, ResourceManager pResourceManager, ProfilerFiller pPreparationsProfiler, ProfilerFiller pReloadProfiler, Executor pBackgroundExecutor, Executor pGameExecutor) {
+        return this.load(pResourceManager, pBackgroundExecutor, pPreparationsProfiler)
+                .thenCompose(pPreparationBarrier::wait)
+                .thenAcceptAsync(values -> this.results = values, pGameExecutor);
+    }
+
+    public void apply() {
+        results.forEach((key, result) -> this.apply((BaseMappedRegistry) registryAccess.registryOrThrow(key), result));
+
+        // Clear the intermediary maps and objects
+        results = null;
+        System.gc();
+    }
+
+    private <T> void apply(BaseMappedRegistry<T> registry, LoadResult<T> result) {
+        registry.dataMaps.clear();
+        result.results().forEach((key, entries) -> registry.dataMaps.put(
+                key, this.buildDataMap(registry, key, (List) entries)));
+    }
+
+    private <T, R> Map<ResourceKey<R>, T> buildDataMap(Registry<R> registry, DataMapType<T, R> attachment, List<DataMapFile<T, R>> entries) {
+        record WithSource<T, R>(T attachment, Either<TagKey<R>, ResourceKey<R>> source) {}
+        final Map<ResourceKey<R>, WithSource<T, R>> result = new IdentityHashMap<>();
+        final BiConsumer<Either<TagKey<R>, ResourceKey<R>>, Consumer<Holder<R>>> valueResolver = (key, cons) -> key.ifLeft(tag -> registry.getTagOrEmpty(tag).forEach(cons)).ifRight(k -> cons.accept(registry.getHolderOrThrow(k)));
+        final DataMapValueMerger<T, R> merger = attachment instanceof AdvancedDataMapType adv ? adv.merger() : DataMapValueMerger.defaultMerger();
+        entries.forEach(entry -> {
+            if (entry.replace()) {
+                result.clear();
+            }
+
+            entry.values().forEach((tKey, value) -> {
+                if (value.isEmpty()) return;
+
+                valueResolver.accept(tKey, holder -> {
+                    final var newValue = value.get().carrier();
+                    final var key = holder.unwrapKey().get();
```

Comment:
Given it will error if it isn't present anyway we may as well be more explicit about that being a potential outcome by using `orElseThrow()`